addons/mrp_subcontracting/models/stock_move.py

```diff
@@ -57,6 +57,40 @@ def _compute_show_details_visible(self):
             move.show_details_visible = True
         return res
 
+    def _set_quantity_done(self, qty):
+        to_set_moves = self
+        for move in self:
+            if move.is_subcontract:
+                # If 'done' quantity is changed through the move, record components as if done through the wizard.
+                move._auto_record_components(qty)
+                to_set_moves -= move
+        if to_set_moves:
+            super(StockMove, to_set_moves)._set_quantity_done(qty)
+
+    def _auto_record_components(self, qty):
+        self.ensure_one()
+        production = self.move_orig_ids.production_id[-1:]
+        if production.product_tracking == 'lot' and not production.lot_producing_id:
+            production.qty_producing = qty
+            production.action_generate_serial()
+            production._set_qty_producing()  # `action_generate_serial` won't `_set_qty_producing`.as its a lot
+            production.subcontracting_record_component()
+        elif production.product_tracking == 'serial' and not production.lot_producing_id:
+            while float_compare(qty, 0, precision_rounding=self.product_uom.rounding) > 0:
+                production.qty_producing = qty
+                production.action_generate_serial()  # `action_generate_serial` will call `_set_qty_producing`.
+                action = production.subcontracting_record_component()
+                backorder_id = action.get('res_id')
+                # If we have no backorder, the qty was fully absorbed.
+                if not backorder_id:
+                    return
+                production = self.env['mrp.production'].browse(backorder_id)
+                qty -= 1
+        else:
+            production.qty_producing = qty
+            production._set_qty_producing()
+            production.subcontracting_record_component()
```

Comment:
If you change and save qty_done multiple times, the number of recorded components can be messed.
For example, I tried first set qty_done > demand, then change it back to qty_done = demand, the number of comsume components didn't change.